### ISSUE TITLE
Python Custom Multiclass Loss Function Doesn't Work

### ISSUE THREAD

**Issue by wagnew3** [Microsoft/LightGBM#1308]
Custom multiclass loss functions in python don't work-- LightGBM doesn't seem to be learning anything.
In the example code below on the iris dataset, I train lightgbm with the default multiclass loss function and get a training accuracy of 97%. When I train lightgbm on a custom loss function, I get a training accuracy of 43% (basically random). However, when I train xgboost with the same custom loss function, I get a training accuracy of 99%.

## Environment info
Operating System: Ubuntu 16.04
CPU: Intel® Core™ i7-5930K
C++/Python/R version: Python 3.5.2

```
import numpy as np
from sklearn import datasets
from sklearn.preprocessing import OneHotEncoder
import xgboost as xgb
import lightgbm as lgb

#custom loss function
def logregobj(preds, train_data):
    labels = train_data.get_label()
    preds=np.reshape(preds, (len(labels), 3))
    labels = OneHotEncoder(n_values=3, sparse=False).fit_transform(labels.reshape(-1, 1))
    grad = (preds - labels)
    hess = 2.0 * preds * (1.0-preds)
    return grad.flatten(), hess.flatten()

# Iris dataset
iris = datasets.load_iris()
X, Ymc = iris.data, iris.target
Y = OneHotEncoder(sparse=False).fit_transform(Ymc.reshape(-1, 1))

#train lightgbm with default multiclass loss function
params={'num_class':len(np.unique(Ymc)), 'objective': 'multiclass', 'verbose':1, 'num_leaves': 63, 'num_iterations': 10, 'silent':1}
lgb_train = lgb.Dataset(X, Ymc)
bst=lgb.train(params, lgb_train, verbose_eval=False)
preds=bst.predict(X)
pred_labels=np.argmax(preds, axis=1)
train_error=np.sum(pred_labels==Ymc)
#accuracy
print('lightgbm train correct %:', train_error/Ymc.shape[0])

#train lightgbm with custom loss function
params={'num_class':len(np.unique(Ymc)), 'objective': 'multiclass', 'verbose':1, 'num_leaves': 63, 'num_iterations': 10, 'silent':1}
lgb_train = lgb.Dataset(X, Ymc)
bst=lgb.train(params, lgb_train, verbose_eval=False, fobj=logregobj)
preds=bst.predict(X)
pred_labels=np.argmax(preds, axis=1)
train_error=np.sum(pred_labels==Ymc)
#accuracy
print('lightgbm custom loss train correct %:', train_error/Ymc.shape[0])

#train xgboost with same custom loss function
dtrain = xgb.DMatrix(X, label=Ymc)
param = {'max_depth': '6', 'objective':'multi:softprob', 'num_class':len(np.unique(Ymc)), 'tree_method':'hist', 'silent':1}
model = xgb.Booster(params, [dtrain])
bst = xgb.train(param, dtrain, num_boost_round=10, obj=logregobj)
preds=bst.predict(dtrain)
pred_labels=np.argmax(preds, axis=1)
train_error=np.sum(pred_labels==Ymc)
#accuracy
print('xgboost custom loss train error %:', train_error/Ymc.shape[0])
```

**Comment by guolinke** [comment]
@wagnew3 thanks, it seems there are some bugs, i will take a look.

**Comment by guolinke** [comment]
@wagnew3 
please try this custom obj:
```
def logregobj(preds, train_data):
    labels = train_data.get_label()
    preds=np.reshape(preds, (len(labels), 3))
    labels = OneHotEncoder(n_values=3, sparse=False).fit_transform(labels.reshape(-1, 1))
    grad = (preds - labels)
    hess = 2.0 * preds * (1.0-preds)
    return grad.flatten("F"), hess.flatten("F")
```

**Comment by guolinke** [comment]
refer to doc here: https://github.com/Microsoft/LightGBM/blob/master/python-package/lightgbm/basic.py#L1497-L1499
It is different with the XGBoost.

**Comment by wagnew3** [comment]
Thank you!

**Comment by rhgrossman** [comment]
for me to get the objective function to work, i had to do the following: 

def logregobj(labels, preds):
    preds = np.reshape(preds, (len(labels), 3), "F")
    preds = softmax(preds, axis=1)
    labels = OneHotEncoder(n_values=3, sparse=False).fit_transform(labels.reshape(-1, 1))
    grad = (preds - labels)
    hess = 2.0 * preds * (1.0 - preds)
    return grad.flatten('F'), hess.flatten('F')

Thanks for the informative question and answer though!
